### PR TITLE
[asl] Small various features

### PR DESCRIPTION
**Pull request by HadrienRenaud**
This PR implements:

 - [x] HadrienRenaud/herdtools7#29
 - [x] HadrienRenaud/herdtools7#31
 - [x] HadrienRenaud/herdtools7#32
 - [x] HadrienRenaud/herdtools7#33
 - [x] HadrienRenaud/herdtools7#30
 - [x] HadrienRenaud/herdtools7#34

### CODE REVIEW

**Comment by HadrienRenaud** [comment]
@maranget Can we merge this one before it becomes too big? 

If not, tell me what I should do to get it merged? I can squash if needs be

File: asllib/Interpreter.ml [hunk 1]
```diff
@@ -485,6 +470,68 @@ module Make (B : Backend.S) (C : Config) = struct
...
+            if i = n then prev
+            else
+              let* v' = B.get_index i v in
+              let* here = in_values v' ty' in
+              loop (succ i) (and' prev here)
+          in
+          loop 0 m_true
+      | T_Named _ -> assert false
+    in
+    B.choice (in_values v ty) (return true) (return false)
```
Comment (maranget) [inline asllib/Interpreter.ml:533]: This function is unused...

File: asllib/tests/asl/required/subtypes-with.asl [hunk 2]
```diff
@@ -0,0 +1,39 @@
+type rec1 of record {
+  a: integer,
+  b: integer,
```
Comment (maranget) [inline asllib/tests/asl/required/subtypes-with.asl:3]: Optional ',' after last field type definition is allowed (not in ASL document)

File: asllib/tests/asl/required/subtypes-with.asl [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+type rec2bis subtypes rec1 with { 
+  c: integer,
+};
+
+func main() => integer
+begin
+  let a = rec2 {
+    a = 1,
+    b = 2,
+    c = 3
```
Comment (maranget) [inline asllib/tests/asl/required/subtypes-with.asl:19]: Optional ',' after the last field definition is _not_ allowed. Perhaps we could allow this optional comma everywhere? Notice that the ASL document does not allow those terminal ','.

Comment (maranget) [reply asllib/tests/asl/required/subtypes-with.asl:19]: Well thinking twice, following the official grammar is the best option. If we do not we encourage syntax that other tools may not accept. So the current parser is fine.

This does not mean not lobbying for a more consistent syntax,. A minor issue though.

**Comment by maranget** [comment]
Some of the  most involved constructs, that is, concatenation of lhs and  neseted bitfields with types would deserve at least one example in regression test suite.

Appart from this minor suggestion (and teh more important `is_val_of_type`  being unused), LGTM.

File: asllib/tests/asl/required/subtypes-with.asl (hunk 3 above)
Comment (HadrienRenaud) [reply asllib/tests/asl/required/subtypes-with.asl:19]: Good catch! I somehow managed not to commit it...

It should be fixed with the last pushed version.

I also integrated your example, see the error message in `asllib/tests/asl.t/run.t`. Do you think I should test more on this feature?

File: asllib/Interpreter.ml (hunk 1 above)
Comment (HadrienRenaud) [reply asllib/Interpreter.ml:538]: Good catch! I somehow managed not to commit it...

It should be fixed with the last pushed version.

I also integrated your example, see the error message in `asllib/tests/asl.t/run.t`. Do you think I should test more on this feature?

**Comment by HadrienRenaud** [comment]
@maranget thank you for your review!

Last commit adds a few tests and a few bug fixes (see especially the bug in `storage.ml`)

I will need to rebase on #651 which will create a lot of conflicts ...

**Comment by maranget** [comment]
> I will need to rebase on #651 which will create a lot of conflicts ...

You mean that we wait for this rebase before merging?

**Comment by HadrienRenaud** [comment]
> > I will need to rebase on #651 which will create a lot of conflicts ...
> 
> You mean that we wait for this rebase before merging?

Depending on which you merge first, I will have to rebase the other one before it can get merged. And in the end, I will have to resolve the same conflicts so it really doesn't really matter for me which way we're doing this.

**Comment by maranget** [comment]
I have rebased on top of master (which has PR #651 merged-in), Tests pass.

**Comment by jalglave** [comment]
Thank you. Please let's merge.
Jade

**Comment by maranget** [comment]
Merged, thanks  @HadrienRenaud and @jalglave.

**Pull request merged**